### ISSUE TITLE
how to use realm in offline (android)

### ISSUE THREAD

**Issue by ledboot** [realm/realm-java#3624]
in some cases,we develop in offline, so i download realm jar ,put it into project.

error:
```
Error:(3, 0) Cause: com/neenbedankt/gradle/androidapt/AndroidAptPlugin
```

**Comment by Zhuinden** [comment]
Realm is distributed as an AAR since 0.88.0.

So just adding `realm.jar` won't work.

I think Gradle's offline mode **should** be able to handle this by default?

**Comment by ledboot** [comment]
I add `android-apt-1.3.jar` `realm-gradle-plugin-2.1.0.jar` `realm-transformer-2.1.0.jar` into a folder `plugin`

in `build.gradle` 
```
buildscript {
    repositories {
        jcenter()

    }
    dependencies {
        classpath 'com.android.tools.build:gradle:2.1.3'
//        classpath "io.realm:realm-gradle-plugin:1.1.0"
        classpath fileTree(dir:'app/plugin',include: ['*.jar'])
    }
}
```

gradle sync an error:
```
Error:(3, 0) Failed to apply plugin [id 'realm-android']
```

**Comment by Zhuinden** [comment]
But as I asked, **doesn't Gradle already handle this if you enable Offline Mode?**

**Comment by kneth** [comment]
I have created a small project in Android Studio, added Realm Java 2.0.2 to the two relevant Gradle files (as outlined in the documentation), and sync the project. After that, I set Gradle offline, and I can build.

**Comment by cmelchior** [comment]
@ledboot 
If Gradle offline mode is not enough, you can create a local maven repository to fetch the files from. This project here shows how it can be accomplished: http://ge.tt/2axpkVf2

**Comment by DoubleRockStar** [comment]
I was trying to use Realm in offline mode. so,, After gradle succeed, I got following files
realm-annotations-2.0.2.jar
realm-annotations-processor-2.0.2.jar
realm-gradle-plugin-2.0.2.jar
realm-transformer-2.0.2.jar
realm-android-library-2.0.2.aar
in C:\Users\user_name\.gradle\caches\modules-2\files-2.1\io.realm

and I moved 3 files (realm-gradle-plugin-2.0.2.jar, realm-transformer-2.0.2.jar, realm-android-library-2.0.2.aar) into libs folder of project and tried to build.

dependencies {
    ... .... ...
    compile fileTree(dir: 'libs', include: ['*.jar'])
    compile(name:'realm-android-library-2.0.2', ext:'aar')
}

build is successfull , but app crashes when it running on MyApplication
public class MyApplication extends Application {

    @Override
    public void onCreate() {
        super.onCreate();

        // Add your initialization code here
        Parse.initialize(new Parse.Configuration.Builder(this)
                .applicationId("E3bTMbxLnMAzuxER2XBSAZWTwcrGe7st95hrxiz1")
                .clientKey("s6lhRSbT5jTXK6thLuK28bdTQAxUEispX50BhUS4")
                .server("https://parseapi.back4app.com/").build()
        );

        Realm.init(this);
        RealmConfiguration config = new RealmConfiguration.Builder().build();
        Realm.setDefaultConfiguration(config);
    }

}

Here is log for crash

                                                                           --------- beginning of crash
11-03 20:11:19.101 21013-21013/com.tahaclay.gifselfiecam E/AndroidRuntime: FATAL EXCEPTION: main
                                                                           Process: com.tahaclay.gifselfiecam, PID: 21013
                                                                           java.lang.NoClassDefFoundError: Failed resolution of: Lcom/getkeepsafe/relinker/ReLinker;
                                                                               at io.realm.internal.RealmCore.loadLibrary(RealmCore.java:59)
                                                                               at io.realm.Realm.init(Realm.java:188)
                                                                               at com.tahaclay.gifselfiecam.MyApplication.onCreate(MyApplication.java:30)
                                                                               at com.android.tools.fd.runtime.BootstrapApplication.onCreate(BootstrapApplication.java:369)
                                                                               at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1024)
                                                                               at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5361)
                                                                               at android.app.ActivityThread.-wrap2(ActivityThread.java)
                                                                               at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1528)
                                                                               at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                               at android.os.Looper.loop(Looper.java:154)
                                                                               at android.app.ActivityThread.main(ActivityThread.java:6077)
                                                                               at java.lang.reflect.Method.invoke(Native Method)
                                                                               at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                                                               at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
11-03 20:11:19.101 21013-21013/com.tahaclay.gifselfiecam E/AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "com.getkeepsafe.relinker.ReLinker" on path: DexPathList[[dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-support-annotations-25.0.0_fe0f2ce472817c00cee651dc7aa866e51fea3fd3-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_9-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_8-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_7-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_6-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_5-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_4-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_3-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_2-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_1-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-slice_0-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-internal_impl-25.0.0_f2beca9028390f0f92590bfb4e86aae21a3eabc0-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-internal_impl-25.0.0_dc17af393c50955d2a2b4b5fb140108b677946e5-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-internal_impl-25.0.0_d0ef789074538894dc10b3e78a281f566afeab2a-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-internal_impl-25.0.0_96e1c0dce32564aaeb1c7df57d9a36bb1ad8d143-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-internal_impl-25.0.0_835f468c2574f209c6e0db617308876acaa590c5-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-intermediates-exploded-aar-realm-android-library-2.0.2_7a4f741008c33378f293bf2c981898e87c1e18e8-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-intermediates-exploded-aar-android-gif-drawable-1.2.2_f6965183ec0c7186993971ca97e06a97f120e591-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-transition-25.0.0_35b81c91c47088f8a53b80bf06c660a9b530c5ac-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-support-vector-drawable-25.0.0_f9d219704b5ee2325cb4594a240b007f4a1f6b2f-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-support-v4-25.0.0_942ececfea90c6c1cdcb3f6884707a441a745dd3-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-support-media-compat-25.0.0_15e652347444588a92257e28b001b5434ba88311-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-support-fragment-25.0.0_13cf4c47ef990315397f6110df59b1c4aeebe57e-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-support-core-utils-25.0.0_411fd30dfb371c54ceaf1d738fcba2b8b88ca317-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-support-core-ui-25.0.0_5bfac28264433b09e5ac162b9808c989a8503c6e-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-support-compat-25.0.0_08dbdced184e964c2b5c4e5db71fc63ed71fa179-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-recyclerview-v7-25.0.0_ec475055be2879c259a76ae75eca373c600db20c-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.androi
11-03 20:11:19.101 21013-21013/com.tahaclay.gifselfiecam E/AndroidRuntime: d.support-design-25.0.0_0ac3a1675a5a3953afc1825c923af3809e10e104-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-appcompat-v7-25.0.0_f4aee45d4fb140efa9d704cd810cfea814733d5f-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-com.android.support-animated-vector-drawable-25.0.0_abf384d112ec15bba687a7cd71460064be700e21-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-bolts-tasks-1.4.0_1ff8976e5aac4575530f62a2286b2a97c8199e69-classes.dex", dex file "/data/data/com.tahaclay.gifselfiecam/files/instant-run/dex/slice-Parse-1.13.0_231e35da4cbcaf0b0437ba259c2a34a510924abe-classes.dex"],nativeLibraryDirectories=[/data/app/com.tahaclay.gifselfiecam-2/lib/x86_64, /data/app/com.tahaclay.gifselfiecam-2/base.apk!/lib/x86_64, /system/lib64, /vendor/lib64]]
                                                                               at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
                                                                               at com.android.tools.fd.runtime.IncrementalClassLoader$DelegateClassLoader.findClass(IncrementalClassLoader.java:90)
                                                                               at java.lang.ClassLoader.loadClass(ClassLoader.java:380)
                                                                               at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
                                                                               	... 14 more

**Comment by kneth** [comment]
Realm depends on the ReLinker library so you need to download that too.

**Comment by diegomontoya** [comment]
@cmelchior  The link http://ge.tt/2axpkVf2 is blocked by chrome and any download is automatically flagged as malicious from the site. Firefox doesn't even trigger a download. Can you post the project to another file share? Thanks.

**Comment by DoubleRockStar** [comment]
@kneth , I added ReLinker library as you told, but another crash issue 

    dependencies {
    ......
         compile(name: 'relinker-1.2.2', ext: 'aar')
         compile(name:'realm-android-library-2.0.2', ext:'aar')
   }


public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        mContainer = container;
        View view = inflater.inflate(R.layout.fragment_animate, container, false);

        listView = (ListView) view.findViewById(R.id.fragment_animate_list);

        **results = realm.where(GSCAnimation.class).findAll();** // crashes on here

        loadingView = (LoadingView) view.findViewById(R.id.electric_fan_view);
        adapter = new AnimateListAdapter(getActivity(), results);
        listView.setAdapter(adapter);

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                selectedID = i;
                adapter.setCurrentID(selectedID);
                adapter.notifyDataSetChanged();
            }
        });

        return view;
    }

E/AndroidRuntime: FATAL EXCEPTION: main
                  Process: com.tahaclay.gifselfiecam, PID: 3395
                  java.lang.IllegalArgumentException: GSCAnimation is not part of the schema for this Realm
                      at io.realm.internal.modules.CompositeMediator.getMediator(CompositeMediator.java:172)
                      at io.realm.internal.modules.CompositeMediator.getTableName(CompositeMediator.java:90)
                      at io.realm.RealmSchema.getTable(RealmSchema.java:296)
                      at io.realm.RealmSchema.getSchemaForClass(RealmSchema.java:316)
                      at io.realm.RealmQuery.<init>(RealmQuery.java:138)
                      at io.realm.RealmQuery.createQuery(RealmQuery.java:87)
                      at io.realm.Realm.where(Realm.java:1214)
                      at com.tahaclay.gifselfiecam.fragments.GSCAnimateFragment.onCreateView(GSCAnimateFragment.java:73)
                      at android.app.Fragment.performCreateView(Fragment.java:2353)
                      at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:995)
                      at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1171)
                      at android.app.BackStackRecord.run(BackStackRecord.java:815)
                      at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1578)
                      at android.app.FragmentManagerImpl$1.run(FragmentManager.java:483)
                      at android.os.Handler.handleCallback(Handler.java:751)
                      at android.os.Handler.dispatchMessage(Handler.java:95)
                      at android.os.Looper.loop(Looper.java:154)
                      at android.app.ActivityThread.main(ActivityThread.java:6077)
                      at java.lang.reflect.Method.invoke(Native Method)
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)

**Comment by kneth** [comment]
It looks like the annotation processor hasn't been executed. You might have to add `@RealmClass` to your model classes.
